ledger-service/db-backend/src/main/scala/com/digitalasset/http/dbbackend/Queries.scala

```diff
@@ -577,26 +675,117 @@ private object OracleQueries extends Queries {
-                  FROM contract
-                  WHERE tpid = $tpid""" // TODO SC AND (""" ++ predicate ++ sql")"
+    import Queries.CompatImplicits.catsReducibleFromFoldable1
+    // % is explicitly reserved by specification as a delimiter
+    val q = sql"""SELECT c.contract_id, key, payload, agreement_text, sd.parties, od.parties
+                  FROM (contract c
+                        JOIN signatories sm ON (c.contract_id = sm.contract_id)
+                        JOIN observers om ON (c.contract_id = om.contract_id))
+                       LEFT JOIN (SELECT contract_id, LISTAGG(party, '%') parties
+                                  FROM signatories GROUP BY contract_id) sd
+                              ON (c.contract_id = sd.contract_id)
+                       LEFT JOIN (SELECT contract_id, LISTAGG(party, '%') parties
+                                  FROM observers GROUP BY contract_id) od
+                              ON (c.contract_id = od.contract_id)
+                  WHERE (""" ++ Fragments.in(fr"sm.party", parties) ++
+      fr" OR " ++ Fragments.in(fr"om.party", parties) ++ sql""")
+                        AND tpid = $tpid AND (""" ++ predicate ++ sql")"
     trackMatchIndices match {
       case MatchedQueryMarker.ByInt => sys.error("TODO websocket Oracle support")
       case MatchedQueryMarker.Unused =>
-        q.query[(String, JsValue, JsValue, Option[String])].map {
-          case (cid, key, payload, agreement) =>
+        q.query[(String, JsValue, JsValue, Option[String], Option[String], Option[String])].map {
+          case (cid, key, payload, agreement, pctSignatories, pctObservers) =>
             DBContract(
               contractId = cid,
               templateId = tpid,
-              key = key,
+              key = key.asJsObject.fields("key"),
               payload = payload,
-              signatories = Vector(),
-              observers = Vector(),
+              signatories = unpct(pctSignatories),
+              observers = unpct(pctObservers),
               agreementText = agreement getOrElse "",
             )
         }
     }
   }
+
+  private[this] def unpct(s: Option[String]) =
+    s.cata(_.split('%').filter(_.nonEmpty).toVector, Vector.empty)
+
+  private[http] override def keyEquality(key: JsValue): Fragment = {
+    import spray.json.DefaultJsonProtocol.JsValueFormat
+    sql"JSON_EQUAL(key, ${toDBContractKey(key)})"
+  }
+
+  private[this] def pathSteps(path: JsonPath): Cord =
+    path.elems.foldMap(_.fold(k => (".\"": Cord) ++ k :- '"', (_: _0.type) => "[0]"))
+
+  // I cannot believe this function exists in 2021
+  // ORA-40454: path expression not a literal
+  private[this] def oraclePathEscape(readyPath: Cord): Fragment = {
+    val s = readyPath.toString
+    assert(
+      !s.startsWith("'") && !s.endsWith("'"),
+      "Oracle JSON query syntax doesn't allow ' at beginning or ending",
+    )
+    Fragment const0 ("'" + s.replace("'", "''") + "'")
+  }
```

Comment:
Reviewers, please take special care to examine this function.

Reply:
Can you provide an example of a query that needs this? I think I’m not quite following which weird restriction we’re trying to work around here.

Reply:
Every location that uses the function needs it. Function calls of the form `json_query(x, $y)` where `y: String` fail with `ORA-40454: path expression not a literal`. In other words, while paths use string _syntax_, you cannot put anything but a literal string in the SQL.

Reply:
Let me rephrase, in which case do we get a value that we need to escape? Daml-LF names cannot have `'` in them so record fields or variants are not an option. What can hit this?

Reply:
`$.p?(@ == "I'm a string")` as produced by https://github.com/digital-asset/daml/pull/9159/files#diff-570f7c791ece7e9dc9ef545dae8b7bb7375560172359ddc5a6ed86477713d6c1R736-R739

Reply:
Makes sense, what is stopping us from relying on doobie’s escaping, e.g., something like `fr"${escapethis}"`?

Reply:
Oh I guess that goes via jdbc to do escaping so it’s not a literal to oracle?

Reply:
Exactly.

Reply:
ugh, alright, don’t have a better idea then. This is very sad :sob: